src/modules/ekf2/EKF/aid_sources/external_vision/ev_vel_control.cpp

```diff
@@ -256,92 +142,56 @@ void Ekf::controlEvVelFusion(const imuSample &imu_sample, const extVisionSample
 	}
 }
 
-bool Ekf::fuseEvVelocity(estimator_aid_source3d_s &aid_src, const extVisionSample &ev_sample)
+void Ekf::stopEvVelFusion()
 {
-	if (ev_sample.vel_frame == VelocityFrame::BODY_FRAME_FRD) {
-
-		VectorState H;
-		estimator_aid_source1d_s current_aid_src;
-		const auto state_vector = _state.vector();
-
-		for (uint8_t index = 0; index <= 2; index++) {
-			current_aid_src.timestamp_sample = aid_src.timestamp_sample;
-
-			if (index == 0) {
-				sym::ComputeEvBodyVelHx(state_vector, &H);
-
-			} else if (index == 1) {
-				sym::ComputeEvBodyVelHy(state_vector, &H);
-
-			} else {
-				sym::ComputeEvBodyVelHz(state_vector, &H);
-			}
-
-			const float innov_var = (H.T() * P * H)(0, 0) + aid_src.observation_variance[index];
-			const float innov = (_R_to_earth.transpose() * _state.vel - Vector3f(aid_src.observation))(index, 0);
-
-			updateAidSourceStatus(current_aid_src,
-					      ev_sample.time_us,				// sample timestamp
-					      aid_src.observation[index],			// observation
-					      aid_src.observation_variance[index],		// observation variance
-					      innov,						// innovation
-					      innov_var,    					// innovation variance
-					      math::max(_params.ev_vel_innov_gate, 1.f));	// innovation gate
-
-			if (!current_aid_src.innovation_rejected) {
-				fuseBodyVelocity(current_aid_src, current_aid_src.innovation_variance, H);
+	if (_control_status.flags.ev_vel) {
+		_control_status.flags.ev_vel = false;
+	}
+}
 
-			}
+void Ekf::fuseLocalFrameVelocity(estimator_aid_source3d_s &aid_src, const uint64_t &timestamp,
+				 const Vector3f &measurement, const Vector3f &measurement_var, const float &innovation_gate)
+{
+	updateAidSourceStatus(aid_src,
+			      timestamp,				// sample timestamp
+			      measurement,				// observation
+			      measurement_var,				// observation variance
+			      _state.vel - measurement,			// innovation
+			      getVelocityVariance() + measurement_var,	// innovation variance
+			      innovation_gate);				// innovation gate
+	fuseVelocity(aid_src);
+}
 
-			aid_src.innovation[index] = current_aid_src.innovation;
-			aid_src.innovation_variance[index] = current_aid_src.innovation_variance;
-			aid_src.test_ratio[index] = current_aid_src.test_ratio;
-			aid_src.fused = current_aid_src.fused;
-			aid_src.innovation_rejected |= current_aid_src.innovation_rejected;
+void Ekf::fuseBodyFrameVelocity(estimator_aid_source3d_s &aid_src, const uint64_t &timestamp,
+				const Vector3f &measurement, const Vector3f &measurement_var, const float &innovation_gate)
+{
+	VectorState H[3];
+	Vector3f innov_var;
+	Vector3f innov = _R_to_earth.transpose() * _state.vel - measurement;
+	sym::ComputeBodyVelH(_state.vector(), P, measurement_var, &innov_var, &H[0], &H[1], &H[2]);
```

Comment:
The innovation changes after each P update, so it needs to be re-computed after fusing the x axis and again after fusing the y axis (but keep the same linearization point).
See section 3.2 (p.30-31) of https://ntrs.nasa.gov/api/citations/20180003657/downloads/20180003657.pdf . We want the sequential fusion to be independent on the processing order of the vector measurement.